File: frontend/src/types/alert.ts
export type Severity = "critical" | "high" | "medium" | "low" | "info";

export interface CategoryBrief {
  id: number;
  name: string;
}

export interface Alert {
  id: number;
  raw_alert_id: number;
  title: string;
  description: string | null;
  severity: Severity;
  cvss_score: string | null; // Decimal serializes as string in JSON
  source_name: string;
  affected_products: Record<string, unknown> | null;
  attack_vectors: Record<string, unknown> | null;
  mitre_techniques: Record<string, unknown> | null;
  iocs: Record<string, unknown> | null;
  published_date: string | null;
  normalized_at: string;
  categories: CategoryBrief[];
  notes: string | null;
  is_acknowledged: boolean;
  acknowledged_at: string | null;
}

export type AlertSortField =
  | "normalized_at"
  | "published_date"
  | "severity"
  | "cvss_score"
  | "source_name"
  | "title";

export interface AlertListResponse {
  items: Alert[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface AlertStatsResponse {
  total_alerts: number;
  by_severity: Record<string, number>;
  by_source: Record<string, number>;
  last_24h: number;
  last_7d: number;
}

export interface AlertFilters {
  page: number;
  page_size: number;
  severity?: string;
  source?: string;
  search?: string;
  category?: string;
  date_from?: string;
  date_to?: string;
  sort_by?: AlertSortField;
  sort_order?: "asc" | "desc";
  is_acknowledged?: boolean;
}

export interface CorrelationGroup {
  group_type: "cve" | "vendor";
  key: string;
  count: number;
  severities: string[];
  sources: string[];
  alert_ids: number[];
}

export interface CorrelationsResponse {
  groups: CorrelationGroup[];
}

File: app/schemas/alert.py
"""Pydantic schemas for alerts."""

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class AlertBase(BaseModel):
    """Base alert fields."""

    title: str
    description: str | None = None
    severity: str = "info"
    cvss_score: Decimal | None = None
    source_name: str


class AlertNotesUpdate(BaseModel):
    """Payload for updating analyst notes on an alert."""
    notes: str | None = Field(default=None, max_length=5000)


class AlertResponse(AlertBase):
    """Alert response with all fields."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    raw_alert_id: int
    affected_products: dict | None = None
    attack_vectors: dict | None = None
    mitre_techniques: dict | None = None
    iocs: dict | None = None
    published_date: datetime | None = None
    normalized_at: datetime
    categories: list["CategoryBrief"] = []
    notes: str | None = None
    is_acknowledged: bool = False
    acknowledged_at: datetime | None = None


class CategoryBrief(BaseModel):
    """Minimal category info for embedding in alert responses."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str


class AlertListResponse(BaseModel):
    """Paginated list of alerts."""

    items: list[AlertResponse]
    total: int
    page: int
    page_size: int
    pages: int


class AlertStatsResponse(BaseModel):
    """Aggregated alert statistics."""

    total_alerts: int
    by_severity: dict[str, int]
    by_source: dict[str, int]
    last_24h: int
    last_7d: int


# Resolve forward reference
AlertResponse.model_rebuild()
